scheduler/core/events.go

```diff
@@ -35,19 +37,69 @@ type event interface {
 }
 
 type state struct {
-	sched       scheduler.Scheduler
-	peerManager daemon.PeerMgr
-	cdnManager  daemon.CDNMgr
+	sched                         scheduler.Scheduler
+	peerManager                   daemon.PeerMgr
+	cdnManager                    daemon.CDNMgr
+	waitScheduleParentPeerQueue   workqueue.DelayingInterface
+	waitScheduleChildrenPeerQueue workqueue.DelayingInterface
 }
 
 func newState(sched scheduler.Scheduler, peerManager daemon.PeerMgr, cdnManager daemon.CDNMgr) *state {
 	return &state{
-		sched:       sched,
-		peerManager: peerManager,
-		cdnManager:  cdnManager,
+		sched:                         sched,
+		peerManager:                   peerManager,
+		cdnManager:                    cdnManager,
+		waitScheduleParentPeerQueue:   workqueue.NewNamedDelayingQueue("wait reSchedule parent"),
+		waitScheduleChildrenPeerQueue: workqueue.NewNamedDelayingQueue("wait reSchedule children"),
 	}
 }
 
+func (s *state) start() {
+	go func() {
+		for {
+			v, shutdown := s.waitScheduleParentPeerQueue.Get()
+			if shutdown {
+				break
+			}
+			peer := v.(*types.Peer)
+			parent, candidates, hashParent := s.sched.ScheduleParent(peer)
+			if !hashParent {
+				logger.WithTaskAndPeerID(peer.Task.TaskID, peer.PeerID).Warnf("waitScheduleParentPeerQueue: there is no available parent，reschedule it in one second")
+				s.waitScheduleParentPeerQueue.AddAfter(peer, time.Second)
+				continue
+			}
+			if peer.PacketChan == nil {
+				logger.Errorf("waitScheduleParentPeerQueue: there is no packet chan associated with peer %s", peer.PeerID)
+				return
+			}
+			peer.PacketChan <- constructSuccessPeerPacket(peer, parent, candidates)
+			s.waitScheduleParentPeerQueue.Done(v)
+		}
+	}()
+	go func() {
```

Comment:
The caller uses `state.start()` with `go state.start()`, the second `go func() {}` in this is unnecessary, it can reuse the goroutine with `go state.start()`.